e2e/nsenter.go

```diff
@@ -0,0 +1,290 @@
+package e2e_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/Azure/agentbaker/pkg/agent/datamodel"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/sanity-io/litter"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/remotecommand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	sshCommandTemplate                  = `echo '%s' > sshkey && chmod 0600 sshkey && ssh -i sshkey -o PasswordAuthentication=no -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o ConnectTimeout=5 azureuser@%s sudo`
+	listVMSSNetworkInterfaceURLTemplate = "https://management.azure.com/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/virtualMachineScaleSets/%s/virtualMachines/%d/networkInterfaces?api-version=2018-10-01"
+)
+
+func extractLogsFromVM(ctx context.Context, t *testing.T, cloud *azureClient, kube *kubeclient, subscription, resourceGroupName, clusterName, vmssName, sshPrivateKey string) (map[string]string, error) {
+	pl := cloud.coreClient.Pipeline()
+	url := fmt.Sprintf(listVMSSNetworkInterfaceURLTemplate,
+		subscription,
+		agentbakerTestResourceGroupName,
+		vmssName,
+		0,
+	)
+	req, err := runtime.NewRequest(ctx, "GET", url)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := pl.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	respBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var instanceNICResult listVMSSVMNetworkInterfaceResult
+
+	if err := json.Unmarshal(respBytes, &instanceNICResult); err != nil {
+		return nil, err
+	}
+
+	privateIP := instanceNICResult.Value[0].Properties.IPConfigurations[0].Properties.PrivateIPAddress
+
+	sshCommand := fmt.Sprintf(sshCommandTemplate, sshPrivateKey, privateIP)
+
+	commandList := map[string]string{
+		"/var/log/azure/cluster-provision.log": "cat /var/log/azure/cluster-provision.log",
+	}
+
+	podList := corev1.PodList{}
+	if err := kube.dynamic.List(context.Background(), &podList, client.MatchingLabels{"app": "debug"}); err != nil {
+		return nil, fmt.Errorf("failed to list debug pod: %q", err)
+	}
+
+	if len(podList.Items) < 1 {
+		return nil, fmt.Errorf("failed to find debug pod, list by selector returned no results")
+	}
+
+	podName := podList.Items[0].ObjectMeta.Name
+
+	var result = map[string]string{}
+	for file, sourceCmd := range commandList {
+		mergedCmd := fmt.Sprintf("%s %s", sshCommand, sourceCmd)
```

Comment:
good point. trying now!

Reply:
I guess I'm confused why we need the SSH command at all, if we can just exec into the pod and nsenter in order to get the files - it seems like right now we just create the debug pod, exec into it, and SSH into the target VM from there since it's within the same private network as the VM

why couldn't we just add a label selector to the debug pod to always guarantee it spawns on the target VM -> exec -> nsenter -> cat, eliminating SSH altogether

Reply:
I think this was in case the node didnt actually join the api server for some reason or cse failed to execute. I think.

Reply:
there are two cases for extraction:
- to get the parameters from an existing (healthy) node - easiest way is nsenter via pod, no ssh
- to extract logs from a failed node - requires ssh or similar, can't be done with nsenter
  - however, nodes only have private IPs, so easiest way is to jumpbox an existing node via a pod

you don't need to nsenter and then ssh for the 2nd case

Reply:
openssh isn't installed by the container by default so this doesn't work